File: os03_thread/time_thread_create.py
import sys
import threading
import time

def cur_time():
    return int(time.time() * 1e9)

def do_nothing():
    # print(threading.current_thread())
    return

def main():
    n = int(sys.argv[1]) if 1 < len(sys.argv) else 5
    t0 = cur_time()

    # print(threading.current_thread())
    for i in range(n):
        child_th = threading.Thread(target=do_nothing, args=())
        child_th.start()
        child_th.join()
        
  
    t1 = cur_time()
    dt = t1 - t0
    print(f"{dt} nsec to thread_create-and-join {n} threads ({dt/n} nsec/thread)")

main()

File: os03_thread/time_thread_create.c
/* 必要な #include を補うこと (man ページを参照) */
#include <err.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <pthread.h>

long cur_time() {
  struct timespec ts[1];
  clock_gettime(CLOCK_REALTIME, ts);
  return ts->tv_sec * 1000000000L + ts->tv_nsec;
}

void * do_nothing(void * arg) {
  /**
  pthread_t my_thread_id = pthread_self();  
  printf("thread_id: %lu\n", my_thread_id);
  **/
  return arg;
}

int main(int argc, char ** argv) {
  int n = (argc > 1 ? atoi(argv[1]) : 5);
  long t0 = cur_time();
  pthread_t child_thread_id;

  for (int i=0;i<n;i++) {    
    if (pthread_create(&child_thread_id, 0, do_nothing, 0)) {
      err(1, "pthread_create");
    }
    void * ret = 0;
    if (pthread_join(child_thread_id, &ret)) {
      err(1, "pthread_join");
    }
  }
  long t1 = cur_time();
  long dt = t1 - t0;
  printf("%ld nsec to pthrea_create-and-join %d threads (%ld nsec/thread)\n",
         dt, n, dt / n);
  return 0;
}
